Unable to show text alternatives when tapping grammatical error in Mail compose

whsieh opened the pull request:
#### e4ff4c63704ce1768b58e633aec0f12381cd3f25
<pre>
Unable to show text alternatives when tapping grammatical error in Mail compose
<a href="https://bugs.webkit.org/show_bug.cgi?id=259008">https://bugs.webkit.org/show_bug.cgi?id=259008</a>
rdar://111404885

Reviewed by Tim Horton and Abrar Rahman Protyasha.

After the changes in 264745@main, we no longer perform sync IPC underneath
`-requestAutocorrectionContextWithCompletionHandler:`, and instead just return an empty context when
out-of-process keyboard is enabled. This is because UIKit now syncs document state by requesting
`UIWKDocumentContext`s when OOPK is enabled, instead of `UIWKAutocorrectionContext`.

However, there&apos;s still one instance where UIKit still depends on the legacy autocorrection context
to drive platform features: the case of showing grammar text checking results when the user taps on
a grammar correction in editable content (i.e. words with blue dotted underlines). Work around this
for the time being by falling back to shipping behavior when grammar corrections are present in the
document; we can remove this once rdar://111936621 is fixed.

* LayoutTests/editing/selection/ios/show-grammar-replacements-on-tap-expected.txt: Added.
* LayoutTests/editing/selection/ios/show-grammar-replacements-on-tap.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:

Add a layout test to exercise the fix, building on the infrastructure added in 265869@main. Note
that we use `runSingly=true` here because otherwise, `UIWKTextInteractionAssistant` might use a
previously initialized and cached `UITextChecker` instance, rather than the `LayoutTestSpellChecker`
that contains the canned text corrections that the test depends on.

* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Add a new bit to `EditorState` here, indicating whether we have grammar markers inside the current
editable root.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView requestAutocorrectionContextWithCompletionHandler:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):

Canonical link: <a href="https://commits.webkit.org/265880@main">https://commits.webkit.org/265880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bed8246ac1660ac927154e5941b041bb27a20793

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12482 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13881 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11710 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12496 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14397 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12299 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10276 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14297 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11017 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11473 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11176 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14351 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11665 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9615 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10881 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10892 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2976 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15207 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11518 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
Nice test! :shipit:

whsieh committed:
Unable to show text alternatives when tapping grammatical error in Mail compose
https://bugs.webkit.org/show_bug.cgi?id=259008
rdar://111404885

Reviewed by Tim Horton and Abrar Rahman Protyasha.

After the changes in 264745@main, we no longer perform sync IPC underneath
`-requestAutocorrectionContextWithCompletionHandler:`, and instead just return an empty context when
out-of-process keyboard is enabled. This is because UIKit now syncs document state by requesting
`UIWKDocumentContext`s when OOPK is enabled, instead of `UIWKAutocorrectionContext`.

However, there's still one instance where UIKit still depends on the legacy autocorrection context
to drive platform features: the case of showing grammar text checking results when the user taps on
a grammar correction in editable content (i.e. words with blue dotted underlines). Work around this
for the time being by falling back to shipping behavior when grammar corrections are present in the
document; we can remove this once rdar://111936621 is fixed.

* LayoutTests/editing/selection/ios/show-grammar-replacements-on-tap-expected.txt: Added.
* LayoutTests/editing/selection/ios/show-grammar-replacements-on-tap.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:

Add a layout test to exercise the fix, building on the infrastructure added in 265869@main. Note
that we use `runSingly=true` here because otherwise, `UIWKTextInteractionAssistant` might use a
previously initialized and cached `UITextChecker` instance, rather than the `LayoutTestSpellChecker`
that contains the canned text corrections that the test depends on.

* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Add a new bit to `EditorState` here, indicating whether we have grammar markers inside the current
editable root.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView requestAutocorrectionContextWithCompletionHandler:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):

Canonical link: https://commits.webkit.org/265880@main
